code/sorting/quick_sort/quick_sort.cpp

```diff
@@ -1,56 +1,134 @@
+
+template<typename _Random_Acccess_Iter, typename _Compare>
+void quicksort(_Random_Acccess_Iter begin, _Random_Acccess_Iter end, _Compare comp);
+
+template<typename _Random_Acccess_Iter>
+void quicksort(_Random_Acccess_Iter begin, _Random_Acccess_Iter end);
+
+// implementation for 'pointer of POD(Plain Old Data structure)'
+// int* explicit specialization;
+template<typename _Compare>
+void quicksort(int *begin, int *end, _Compare comp);
+
+void quicksort(int *begin, int *end);
+*/
+
+#include <functional>
+
+// for test
+// #include <iostream>
+// #include <algorithm>
+// #include <iomanip>
+// #include <vector>
+// #include <deque>
+// #include <array>
+// #include <string>
+
+// implementation for 'standard random access container'
+template<typename _Random_Acccess_Iter, typename _Compare>
+void quicksort_impl(_Random_Acccess_Iter first, _Random_Acccess_Iter last, _Compare comp) {
+    if (first < last) {
+        // first is pivot
+        _Random_Acccess_Iter i = first, j = last + 1;
+        while (true) {
+            while (i + 1 <= last && comp(*++i, *first))
+                ;
+            while (j - 1 >= first && comp(*first, *--j))
+                ;
+            if (i >= j)
+                break;
+            std::swap(*i, *j);
         }
+        std::swap(*first, *j);
+        quicksort_impl(first, j - 1, comp);
+        quicksort_impl(j + 1, last, comp);
     }
-    swap(v[i + 1], v[hi]);
-    return (i + 1);
 }
 
-void quicksort(vector<int> &v, int lo, int hi)
-{
-    if (lo < hi)
-    {
-        int pi = partition(v, lo, hi);
-        // Separately sort elements
-        quicksort(v, lo, pi - 1);
-        quicksort(v, pi + 1, hi);
-    } 
+template<typename _Random_Acccess_Iter, typename _Compare>
+void quicksort(_Random_Acccess_Iter begin, _Random_Acccess_Iter end, _Compare comp) {
+    if (begin == end)
+        return;
+    --end;
+
+    return quicksort_impl(begin, end, comp);
 }
 
-void print(vector<int> &v)
-{
-    for(int i = 0; i < v.size(); ++i)
-    {
-        cout << v[i] << " ";
-    }
-    cout << endl;
+template<typename _Random_Acccess_Iter>
+void quicksort(_Random_Acccess_Iter begin, _Random_Acccess_Iter end) {
+    return quicksort(begin, end, std::less<typename _Random_Acccess_Iter::value_type>());
 }
 
-// Testing the quick sort implementation
-int main()
-{
-    vector<int> v;
-    for (int i = 1; i <= 10; i++)
-      v.push_back(i);
-    random_shuffle(v.begin(), v.end());
-
-    print(v);
-    quicksort(v, 0, v.size() - 1);
-    print(v);
-    return 0;
+// implementation for 'pointer of POD(Plain Old Data structure)'
+// int* explicit specialization;
+template<typename _Compare>
+void quicksort(int *begin, int *end, _Compare comp) {
+    if (begin == end)
+        return;
+    --end;
+
+    return quicksort_impl(begin, end, comp);
 }
+
+void quicksort(int *begin, int *end) {
+    return quicksort(begin, end, std::less<int>());
+}
+
+//// for test
```

Comment:
Kindly use multiline comments like `/* multiple lines */` It will be easy for an user to uncomment and test 👍